src/tlo/methods/collect_event_chains.py

```diff
@@ -0,0 +1,370 @@
+    
+        # This is how I am passing data from fnc taking place before event to the one after
+        # It doesn't seem very elegant but not sure how else to go about it
+        self.print_chains = False
+        self.df_before = []
+        self.row_before = pd.Series()
+        self.mni_instances_before = False
+        self.mni_row_before = {}
+        self.entire_mni_before = {}
+        
+    PARAMETERS = {
+        # Options within module
+        "generate_event_chains": Parameter(
+            Types.BOOL, "Whether or not we want to collect chains of events for individuals"
+        ),
+        "modules_of_interest": Parameter(
+            Types.LIST, "Restrict the events collected to specific modules. If *, print for all modules"
+        ),
+        "events_to_ignore": Parameter(
+            Types.LIST, "Events to be ignored when collecting chains"
+        ),
+        }
+        
+    def initialise_simulation(self, sim):
+        notifier.add_listener("simulation.pop_has_been_initialised", self.on_notification_pop_has_been_initialised)
+        notifier.add_listener("simulation.on_birth", self.on_notification_of_birth)
+        notifier.add_listener("event.about_to_run", self.on_notification_event_about_to_run)
+        notifier.add_listener("event.has_just_ran", self.on_notification_event_has_just_ran)
+        
+    def read_parameters(self, resourcefilepath: Optional[Path] = None):
+        self.load_parameters_from_dataframe(pd.read_csv(resourcefilepath/"ResourceFile_GenerateEventChains/parameter_values.csv"))
+        
+    def initialise_population(self, population):
+        # Use parameter file values by default, if not overwritten
+        self.generate_event_chains = self.parameters['generate_event_chains'] \
+            if self.generate_event_chains is None \
+            else self.generate_event_chains
+            
+        self.modules_of_interest = self.parameters['modules_of_interest'] \
+            if self.modules_of_interest is None \
+            else self.modules_of_interest
+            
+        self.events_to_ignore = self.parameters['events_to_ignore'] \
+            if self.events_to_ignore is None \
+            else self.events_to_ignore
+            
+        # If modules of interest is '*', set by default to all modules included in the simulation
+        if self.modules_of_interest == ['*']:
+            self.modules_of_interest = list(self.sim.modules.keys())
+
+    def on_birth(self, mother, child):
+        # Could the notification of birth simply take place here?
+        pass
+        
+    def on_notification_pop_has_been_initialised(self, data):
+
+        # When logging events for each individual to reconstruct chains,
+        # only the changes in individual properties will be logged.
+        # At the start of the simulation + when a new individual is born,
+        # we therefore want to store all of their properties
+        # at the start.
+        if self.generate_event_chains:
```

Comment:
I'd always start these with the negating condition, and exit. Then you get an unindented block of code for the interesting logic. i.e. 

```
if not self.generate_event_chains:
    return

ednav = ...
```

`ednav` is not a great variable name (EAV is a well-known structure for data)